proposals/0002-spatial-navigation.md

```diff
@@ -0,0 +1,143 @@
+
+## Code and UI changes
+
+This new feature will follow these steps:
+- Introduce a public option named `enableRCUNavigation`.
+- When a player is set up with the rcu navigation feature active, initiate the SpatialNavigation instance.
+- Collect every focusable component and incorporate them into the SpatialNavigation instance.
+- Focus first available navigable candidate.
+- Once an arrow is pressed we should go through Basic Spatial Navigation Heuristics (call navigate(direction))
+  - Find Candidates
+  - Select the best candidate
+  - Focus the best candidate
+
+### In terms of changes to the code:
+
+We need to make sure that events can still move upwards when spatial navigation is turned on (`enableRCUNavigation` is set to true). Currently, it’s blocking all keydown events with `event.stopPropagation()`, except for TAB key.
+Block the UP and DOWN keys to execute `stepForward()` and `stepBack()` functions, as these keys will be needed for navigation from and to the progress bar (Slider).
+
+### In terms of adding the code
+
+Spatial-navigation interface and Component class extension
+
+```
+interface Positions {
+boundingClientRect: DOMRectReadOnly;
+center: DOMRectReadOnly;
+}
+
+interface Component {
+// Is considered as focusable component.
+getIsFocusable(): boolean;
+// Is currently visible/enabled/etc...
+getIsAvailableToBeFocused(): boolean;
+// on focus handler
+handleFocus(): void;
+// on blur handler
+handleBlur(): void;
+// on submit handler
+handleEnter(): void;
+// component's element boundingClientRect and center
+getPositions(): Positions;
+// focus component's element
+focus(): void;
+}
+
+enum SpatialNavigationDirection {
+Left = 'left',
+Right = 'right',
+Up = 'up',
+Down = 'down'
+}
+
+interface SpatialNavigation {
+// current focusable components:
+getComponents(): Set<Component>;
+// start list keydown events
+start(): void;
+// stop listen key down events
+stop(): void;
+// add focusable component
+add(component: Component): void;
+// remove focusable component
+remove(component: Component): void;
+// temporary pause spatial navigation
+pause(): void;
+// resume spatial navigation
+resume(): void;
+// clear current list of focusable components
+clear(): void;
+// run spatial navigation Heuristics (re-use https://github.com/WICG/spatial-navigation/blob/main/polyfill/spatial-navigation-polyfill.js#L147)
+move(direction: SpatialNavigationDirection): void;
+}
+```
+
+### In terms of changes the UI
+- Since LEFT and RIGHT keys are reserved for video seeking, we’ll have to change the UI and move the progress bar on top of the buttons in order to allow the users to use those keys to navigate between elements when buttons are focused and seek the video when the progress bar is focused
+- Current player controls are too small for big screens, so we’re aiming to make those bigger
```

Comment:
Agreed. We'll tackle that outside this feature, but good to call it out! 👍

Reply:
OK, removing it from the RFC since it's out of scope.

Reply:
Ok, cool. I like that idea. So, to make sure it's understood, the two options would be:
```
spatialNavigation: {
  enabled: true,
  seek: 'horizontal'
}
```
This would be used when you position the progress bar above the control bar - as in the Brightcove Player.
```
spatialNavigation: {
  enabled: true,
  seek: 'vertical'
}
```
This would be used when you position the progress bar in its default position within the control bar - as in Video.js.

Does this mean `vertical` would be the default?

Food for thought: if there are only two possible values for `seek`, should it instead be another boolean, maybe `horizontalSeek: [true|false]`? At a minimum, I'd suggest renaming it to `seekDirection: '[horizontal|vertical]'` to be a bit more semantically meaningful.

Reply:
I like your idea @misteroneill to be boolean, makes more sense.